src/Thorn_Calculus-Core_Bisimilarities.thy

```diff
@@ -1350,9 +1350,148 @@ text \<open>
+    have "A \<triangleright>\<^sup>\<infinity> x. \<P> x \<rightarrow>\<^sub>s\<lparr>A \<triangleright> \<star>\<^bsup>n\<^esup> X\<rparr> post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)" and \<open>\<alpha> = A \<triangleright> \<star>\<^bsup>n\<^esup> X\<close>
+      by (simp_all only:)
+    moreover
+    have "
+      (A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n \<parallel> post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)
+      \<sim>\<^sub>s
+      post_receive n X \<P> \<parallel> (A \<triangleright>\<^sup>\<infinity> x. \<P> x \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n"
+    proof -
+      have "
+        post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)
+        =
+        post_receive n X \<P> \<parallel> (A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n"
+        using post_receive_after_parallel
+        by simp
+      then have "
+        (A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n \<parallel> post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)
+        \<sim>\<^sub>s
+        post_receive n X \<P> \<parallel> ((A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n \<parallel> (A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n)"
+        by (simp only: parallel_left_commutativity)
+      then show ?thesis
+        by (simp only: adapted_after_parallel)
+    qed
+    ultimately show ?thesis
+      unfolding
+        \<open>\<alpha> = A \<triangleright> \<star>\<^bsup>n\<^esup> X\<close> and \<open>Q' = post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)\<close> and
+        \<open>S = (A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n \<parallel> Q'\<close>
+      using
+        post_receive_after_parallel and
+        composition_in_universe [
+          OF
+            suffix_adapted_mutation_in_universe
+            parallel_mutation_in_universe
+        ]
+      by (intro exI conjI, use in assumption) (fastforce intro: rev_bexI)
+  qed (blast elim: repeated_receive_evolution(1))+
+next
+  case (backward_simulation \<alpha> S)
+  then show ?case
+  proof (cases rule: repeated_receive_evolution [case_names receiving])
+    case (receiving n X)
+    from receiving and \<open>A \<triangleright>\<^sup>\<infinity> x. \<P> x \<rightarrow>\<^sub>s\<lparr>\<alpha>\<rparr> S\<close>
+    have "A \<triangleright>\<^sup>\<infinity> x. \<P> x \<rightarrow>\<^sub>s\<lparr>A \<triangleright> \<star>\<^bsup>n\<^esup> X\<rparr> post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)"
+      by (simp only:)
```

Comment:
Interestingly, here you go directly from the given transition to the more specific formulation of this transition. This might be a good approach also for the `forward_simulation` case and is an interesting alternative to showing the underlying equations `η = Receiving` and `A = A'`, which I suggested above.

Reply:
Though feasible, applying a similar strategy for the `forward_simulation` case is a bit more involved, since facts such as `α = A ▹ ⋆⇗n⇖ X` are given as assumptions in the `receiving` subcase of the `backward_simulation` case, whereas, on the other hand, such facts need to be proven in the `forward_simulation` case. As an example, the following is an alternative proof, in the spirit of the mentioned strategy, for the `parallel_left_io` case:
```isabelle
    case (parallel_left_io η A' n X P')
    from ‹A ▹⇧∞ x. 𝒫 x →⇩s⦇IO η A' n X⦈ P'› and ‹α = IO η A' n X›
    have
      "A ▹⇧∞ x. 𝒫 x →⇩s⦇A ▹ ⋆⇗n⇖ X⦈ post_receive n X (λx. 𝒫 x ∥ A ▹⇧∞ x. 𝒫 x)"
    and
      "α = A ▹ ⋆⇗n⇖ X"
    and
      "P' = post_receive n X (λx. 𝒫 x ∥ A ▹⇧∞ x. 𝒫 x)"
      unfolding atomize_conj
      by (-, fastforce elim: transition_from_repeated_receive)
    moreover
    have "
      post_receive n X (λx. 𝒫 x ∥ A ▹⇧∞ x. 𝒫 x) ∥ (A ▹⇧∞ x. 𝒫 x) « suffix n
      ∼⇩s
      post_receive n X 𝒫 ∥ (A ▹⇧∞ x. 𝒫 x ∥ A ▹⇧∞ x. 𝒫 x) « suffix n"
      (* same proof as current *)
    ultimately show ?thesis
      unfolding
        ‹α = A ▹ ⋆⇗n⇖ X› and
        ‹P' = post_receive n X (λx. 𝒫 x ∥ A ▹⇧∞ x. 𝒫 x)› and ‹S = P' ∥ (A ▹⇧∞ x. 𝒫 x) « suffix n› and
        post_receive_after_parallel
      using
        composition_in_universe [
          OF
            suffix_adapted_mutation_in_universe
            parallel_mutation_in_universe
        ]
      by (intro exI conjI, use in assumption) (fastforce intro: rev_bexI)
```
Compared with the current proof, this alternative proof needs one less proof step. Would you be satisfied with it?

Reply:
Update: Given the improvement introduced by 4c711e10243ea6a6b7394c40245a635f3efe9a85 in which facts such as `α = A ▹ ⋆⇗n⇖ X` now need to be proved, I think that the alternative proof above would be even more appealing also for the sake of consistency.